packages/flutter/lib/src/rendering/sliver.dart

```diff
@@ -196,7 +196,22 @@ class SliverConstraints extends Constraints {
   /// the earliest visible part of this sliver in the [AxisDirection].
   ///
   /// For example, if [AxisDirection] is [AxisDirection.down], then this is the
-  /// scroll offset at the top of the visible portion of the sliver.
+  /// amount the top of the visible portion of the sliver has been scrolled
+  /// past the top of the viewport.
+  ///
+  /// This value is typically used to compute whether this sliver should still
+  /// protrude into the viewport via [SliverGeometry.paintExtent] and
+  /// [SliverGeometry.layoutExtent] considering how far the beginning of the
+  /// sliver is above the beginning of the viewport.
+  ///
+  /// For slivers past the end of the viewport, viewports typically stops trying
+  /// to compute the [scrollOffset]s of slivers that are in the 'future'
+  /// direction past the currently visible portion of the viewports. Therefore,
+  /// the [scrollOffset] is typically 0 for slivers below the viewport when
+  /// [AxisDirection] is [AxisDirection.down].
+  /// [SliverConstraints.remainingPaintExtent] is typically used to accomplish
+  /// the same goal of computing whether scrolled out slivers should still
+  /// partially 'protrude in' from the bottom of the viewport.
```

Comment:
this last sentence seems wrong? i don't think i've ever seen that, so it's at least not typical, and may in fact not be possible. :-)

Reply:
Maybe I'm saying it wrong. I meant to draw an analogy with scrollOffset to figure out whether you have a layoutExtent between 0 and your height or not. 

If you have a scrollOffset smaller than your height, you might poke out from the top and have a layoutExtent.

Conversely, if you have a remainingPaintExtent (smaller than your height), you might poke out from the bottom and have a layoutExtent.